lib/min_heap.rb

```diff
@@ -44,31 +52,61 @@ def to_s
   end
 
   # This method returns true if the heap is empty
-  # Time complexity: ?
-  # Space complexity: ?
+  # Time complexity: O(1)
+  # Space complexity: O(1)
   def empty?
-    raise NotImplementedError, "Method not implemented yet..."
+    
+    if @store.nil? || @store.length == 0
+      return true
+    else 
+      return false
+    end
+
   end
 
   private
 
   # This helper method takes an index and
   #  moves it up the heap, if it is less than it's parent node.
   #  It could be **very** helpful for the add method.
-  # Time complexity: ?
-  # Space complexity: ?
+  # Time complexity: O(m) where m is height
+  # Space complexity: O(1), no persistent data structure is made beyond 1 temporary reassigned variable
   def heap_up(index)
+    return if index == 0
+
+    if @store[index].key < @store[((index - (index % 3)) / 2)].key 
+      temp = @store[index] 
+      @store[index] = @store[((index - (index % 3)) / 2)]
+      @store[((index - (index % 3)) / 2)] = temp
+    end
+
+    heap_up(((index - (index % 3)) / 2))
     
   end
 
   # This helper method takes an index and 
   #  moves it up the heap if it's smaller
   #  than it's parent node.
+  # left_child = i * 2 + 1 
+  # right_child = i * 2 + 2
   def heap_down(index)
-    raise NotImplementedError, "Method not implemented yet..."
+    return if @store[index].nil?
+
+    if !@store[index * 2 + 1].nil? && @store[index].key > @store[index * 2 + 1].key
+      swap(index, (index * 2 + 1))
+    end
+
+    if !@store[index * 2 + 2].nil? && @store[index].key > @store[index * 2 + 2].key
+      swap(index, (index * 2 + 2))
+    end
+
+    heap_down(index * 2 + 1)
+    heap_down(index * 2 + 2)
```

Comment:
You're doing a bunch of extra swaps here.  Instead find the smallest child and then if it's smaller than the parent, swap those and continue.  You should never need to swap both the right and left children, as well as heap down both children.